File: src/games/stendhal/client/events/ProgressStatusEvent.java
/* $Id$ */
/***************************************************************************
 *                   (C) Copyright 2003-2010 - Stendhal                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.client.events;

import org.apache.log4j.Logger;

import games.stendhal.client.entity.RPEntity;
import games.stendhal.client.gui.progress.ProgressLogController;

/**
 * adjust the quest progress view
 *
 * @author hendrik
 */
class ProgressStatusEvent extends Event<RPEntity> {
	private static Logger logger = Logger.getLogger(ProgressStatusEvent.class);

	/**
	 * executes the event
	 */
	@Override
	public void execute() {
		try {
			if (!event.has("progress_type")) {
				ProgressLogController.get().showCategories(event.getList("data"));
			} else if (!event.has("item")) {
				if (event.get("progress_type").equals("repeatable")) {
					ProgressLogController.get().setRepeatable(event.getList("data"));
				} else {
					ProgressLogController.get().showCategorySummary(event.get("progress_type"), event.getList("data"));
				}
			} else {
				ProgressLogController.get().showDescription(event.get("progress_type"), event.get("item"), event.get("description"), event.get("information"), event.getList("data"));
			}
		} catch (RuntimeException e) {
			logger.error("Failed to process progress status. Event: " + event, e);
		}
	}

}

File: src/js/stendhal/event/ProgressStatusEvent.ts
/***************************************************************************
 *                   (C) Copyright 2005-2024 - Stendhal                    *
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { FloatingWindow } from "../ui/toolkit/FloatingWindow";
import { TravelLogDialog } from "../ui/dialog/TravelLogDialog";
import { ui } from "../ui/UI";
import { UIComponentEnum } from "../ui/UIComponentEnum";
import { RPEvent } from "marauroa"
import { stendhal } from "stendhal";


/**
 * show travel log and details
 */
export class ProgressStatusEvent extends RPEvent {
	public progress_type!: string;
	public data!: string;
	public item!: string;
	public description!: string;

	public execute(_entity: any): void {
		let progressType = this["progress_type"];
		let dataItems = this["data"].substring(1, this["data"].length - 1).split(/\t/);

		let travelLogDialog = ui.get(UIComponentEnum.TravelLogDialog) as TravelLogDialog;
		if (!this["progress_type"]) {
			if (!travelLogDialog) {
				const dstate = stendhal.config.getWindowState("travel-log");
				travelLogDialog = new TravelLogDialog(dataItems);
				new FloatingWindow("Travel Log", travelLogDialog, dstate.x, dstate.y).setId("travel-log");
			} else {
				travelLogDialog.setDataItems(dataItems);
			}
			travelLogDialog.updateTabs();
			return;
		}

		if (!travelLogDialog) {
			return;
		}
		travelLogDialog.updateTabs();

		if (!this["item"]) {
			if (progressType === "repeatable") {
				travelLogDialog.setRepeatable(dataItems);
			} else {
				travelLogDialog.progressTypeData(progressType, dataItems);
			}
		} else {
			travelLogDialog.itemData(progressType, this["item"], this["description"], dataItems);
		}
	}

};
